### PR TITLE
Safe/assert

### PR DESCRIPTION
**Pull request by robdockins**
Implement the UI elements from #284

### CODE REVIEW

**Comment by robdockins** [comment]
I would like to add some test cases, but otherwise I think this PR implements everything I wanted to close #284.

**Comment by brianhuffman** [comment]
The `assert` combinator is a nice addition; this will definitely be useful. Did you put much thought into the argument order?

File: src/Cryptol/Prelude.hs [hunk 1]
```diff
@@ -32,3 +32,4 @@ arrayContents = B.pack [there|lib/Array.cry|]
 
 cryptolTcContents :: String
 cryptolTcContents = [there|lib/CryptolTC.z3|]
+
```
Comment (brianhuffman) [inline src/Cryptol/Prelude.hs:35]: I'm guessing this is one of those make-a-trivial-change-to-get-cabal-to-recompile situations?

**Comment by brianhuffman** [comment]
We should definitely have some regression tests to show how the `~> ERROR` counterexamples are supposed to work, not just for `:prove` but also for `:check`/`:exhaust`. (Also tests to demonstrate `:safe`, of course.)

Did you want any help coming up with examples?

File: lib/Cryptol.cry [hunk 2]
```diff
@@ -866,6 +866,14 @@ primitive error : {a, len} (fin len) => String len -> a
 undefined : {a} a
 undefined = error "undefined"
 
+/**
+ * Assert that the given condition holds, and raise an error
+ * with the given message if it does not.
+ */
+assert : {a, len} (fin len) => Bit -> String len -> a -> a
+assert pred msg x = if pred then x else error msg
+
```
[the remark this answers is not in the document]
Comment (brianhuffman) [reply lib/Cryptol.cry:875]: If we switch `assert` from `len` to `n`, we should probably also change `error` as well:
```
Cryptol> :? error

    error : {a, len} (fin len) => String len -> a
```
It looks like `error` needs a docstring as well. I'll add a ticket for that.

File: src/Cryptol/Prelude.hs (hunk 1 above)
Comment (robdockins) [reply src/Cryptol/Prelude.hs:35]: yup

**Comment by robdockins** [comment]
I haven't thought much about the argument order for `assert`.  Would you prefer to see the message first?

File: lib/Cryptol.cry (hunk 2 above)
[the remark this answers is not in the document]
Comment (robdockins) [reply lib/Cryptol.cry:875]: Sure, I don't have a strong opinion

**Comment by robdockins** [comment]
Help with examples would be nice.  I can (and probably will) come up with some of my own, but having other people bang on it a little would probably be good to see if there are any bugs to shake out.

**Comment by brianhuffman** [comment]
I'll build this branch and try out a few examples with `:prove` and `:exhaust`. I'll pick a few good test cases and share them here.

As for `assert`: I guess in mental version of `assert` I imagined the string argument coming first, but I didn't put any real thought into the argument order. In general, good reasons for choosing a particular argument order might be:
* if there's an established standard that people are used to
* usefulness of partial applications (neither string-only or predicate-only seem too useful to me)
* if one argument is often a large expression, it's nice to have the shorter one first (I expect both the string and predicate will often be fairly small)
* mnemonic value or if some order "reads well" grammatically

So I guess I don't really have any good reasons to favor string-first, and I suppose that assertions read a bit more naturally if we put the predicate right next to the `assert`, like `assert (x < 10) "out of range"` as opposed to `assert "out of range" (x < 5)`. So I'm happy to keep the predicate-first ordering.

**Pull request merged**